lib/charms/parca_k8s/v0/parca_scrape.py

```diff
@@ -705,143 +642,68 @@ def __init__(
             return
 
-        for relation in self._charm.model.relations[self._relation_name]:
-            relation.data[self._charm.app]["scrape_metadata"] = json.dumps(self._scrape_metadata)
-            relation.data[self._charm.app]["scrape_jobs"] = json.dumps(self._scrape_jobs)
+        jobs = [_sanitize_scrape_configuration(job) for job in jobs] or self._jobs or [DEFAULT_JOB]
+        self._publish_app_data(charm.app, relations, JujuTopology.from_charm(charm).as_dict(), jobs)
 
-    def set_scrape_job_spec(self):
-        """Ensure the scrape target information (as passed to this object on __init__) is published.
+    @staticmethod
+    def _publish_app_data(app: ops.Application,
+                          relations: List[ops.Relation],
+                          scrape_metadata: dict,
+                          jobs: List[ScrapeJobsConfig]):
 
-        When a profiling provider charm is related to a Parca charm, the profiling provider sets
-        specification and metadata related to its own scrape configuration. This information is set
-        using Juju application data. Each of the consumer units also sets its own host address in
-        Juju unit relation data.
-        """
-        self._publish_all_relation_data()
+        for relation in relations:
+            relation.data[app]["scrape_metadata"] = json.dumps(scrape_metadata)
+            relation.data[app]["scrape_jobs"] = json.dumps(jobs)
 
-    def _set_unit_ip(self, _event=None):
+    @staticmethod
+    def _publish_unit_data(unit: ops.Unit, relations: List[ops.Relation]):
         """Set unit host address.
 
         Each time a profiling provider charm container is restarted it updates its own host address
         in the unit relation data for the Parca charm. The only argument specified is an event and
         it is ignored.
         """
-        for relation in self._charm.model.relations[self._relation_name]:
-            relation.data[self._charm.unit]["parca_scrape_unit_address"] = socket.getfqdn()
-            relation.data[self._charm.unit]["parca_scrape_unit_name"] = str(
-                self._charm.model.unit.name
-            )
+        for relation in relations:
+            relation.data[unit]["parca_scrape_unit_address"] = socket.getfqdn()
+            relation.data[unit]["parca_scrape_unit_name"] = str(unit.name)
 
-    def _is_valid_unit_address(self, address: str) -> bool:
-        """Validate a unit address.
-
-        Args:
-            address: a string representing a unit address
-        """
-        try:
-            _ = ipaddress.ip_address(address)
-            return True
-        except ValueError:
-            return False
-
-    @property
-    def _relations(self) -> List[Relation]:
-        """The relations currently active on this endpoint."""
-        return self._charm.model.relations[self._relation_name]
-
-    def is_ready(self, relation: Optional[Relation] = None) -> bool:
+    def is_ready(self, charm:ops.CharmBase) -> bool:
```

Comment:
Any reason why the provider has `is_ready` but the consumer does not?  The consumer feels like a more natural fit (its good to have a nice way to know if your charm should be in WaitingStatus because you haven't heard back from a relation).  I don't see a clean way to simulate `is_ready` in the charm without digging into the relations and deciding if they're incomplete